packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -38,27 +90,177 @@ contract FundingRateStore is FundingRateStoreInterface, Testable {
+
+            emit PublishedRate(
+                identifier,
+                fundingRateRecord.rate.rawValue,
+                fundingRateRecord.proposal.time,
+                fundingRateRecord.proposal.proposer
+            );
+        }
+
+        // Compute final fee at time of proposal.
+        FixedPoint.Unsigned memory finalFeeBond = _computeFinalFees();
+
+        uint256 currentTime = getCurrentTime();
+        fundingRateRecord.proposal = Proposal({
+            rate: rate,
+            time: currentTime,
+            proposer: msg.sender,
+            finalFee: finalFeeBond,
+            disputer: address(0x0)
+        });
+        emit ProposedRate(identifier, rate.rawValue, currentTime, msg.sender);
+
+        // Pull final fee bond from proposer.
+        collateralCurrency.safeTransferFrom(msg.sender, address(this), finalFeeBond.rawValue);
+    }
+
+    /**
+     * @notice Dispute a pending funding rate. This will delete the pending funding rate, meaning that a
+     * proposer can now proposer another rate with a fresh liveness.
+     * @dev This will revert if there is no pending funding rate for the identifier.
+     * @dev Caller must approve this this contract to spend `finalFeeBond` amount of collateral, which they can
+     * receive back if their dispute is successful.
+     */
+    function dispute(bytes32 identifier) external {
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(identifier);
+        ProposalState proposalState = _getProposalState(fundingRateRecord.proposal);
+
+        require(proposalState == ProposalState.Pending, "Existing proposal not pending.");
+
+        // TODO: Bond logic.
+
+        // Pull final fee bond from disputer and pay store.
+        _payFinalFees(msg.sender, fundingRateRecord.proposal.finalFee);
+
+        // Send price request
+        _requestOraclePrice(identifier, fundingRateRecord.proposal.time);
+
+        emit DisputedRate(
+            identifier,
+            fundingRateRecord.proposal.rate.rawValue,
+            fundingRateRecord.proposal.time,
+            fundingRateRecord.proposal.proposer,
+            msg.sender
+        );
+
+        // Delete pending proposal and copy into dispute records.
+        fundingRateRecord.proposal.disputer = msg.sender;
+        fundingRateDisputes[identifier][fundingRateRecord.proposal.time] = fundingRateRecord;
+        delete fundingRateRecords[identifier].proposal;
+    }
+
+    /**
+     * @notice Settle a disputed funding rate. The winner of the dispute, either the disputer or the proposer,
+     * will receive a reward plus their final fee bond. This method will also overwrite the current funding rate
+     * with the resolved funding rate returned by the Oracle. Pending funding rates are unaffected by this method.
+     * @dev This will revert if there is no price available for the disputed funding rate.
+     */
+    function settleDispute(bytes32 identifier, uint256 proposalTime) external {
```

Comment:
@nicholaspai, I like this a lot.

There is no `msg.sender` in this method. That's super cool. Something we could do in the perp is have the liquidation withdrawal work the same way -- where all the payouts happen on one call without caring who's calling. That would actually reduce some of the weird edge cases and state we have to maintain.

Thoughts?

Reply:
I was thinking the same thing as we write this. Basically the one who profits the most is incentivized probably to call this, which distributes all of the rewards. I don't actually know why we chose to do it in 3 separate payments before. Maybe we implement this in the PerpetualLiquidarable contract?

Reply:
+1 yeah, we should add to liquidatable. Probably will reduce the bytecode tbh.

Reply:
Created #2149 to track this

Reply:
Nice +1